### ISSUE TITLE
Input Text will POST weird file string into input box in case of Selenium Server

### ISSUE THREAD

**Issue by Perza-e** [robotframework/SeleniumLibrary#1398]
I have a variable ${input_text}= kissa in my robot file. I also have a file called "kissa" within the same folder with the robot file. So when I run the robot script, it will open a browser via Selenium Server, navigates to a web page and inputs a value to an input box. However, it will input a weird string into the input box:

SeleniumLibrary.Input Text is POSTing weird string to the Selenium Server:
![image](https://user-images.githubusercontent.com/52034228/59842378-21d08280-935f-11e9-8ac6-02125d881049.png)

Full text of the weird string:
C:\Users\fipeesk\AppData\Local\Temp\b32956bd1c73e1dd00605472ff462059\upload7217265257046552210file\kissa

**File System:**

![image](https://user-images.githubusercontent.com/52034228/59842066-5f80db80-935e-11e9-9e04-3760e383ad98.png)

**Contents of robot file:**
![image](https://user-images.githubusercontent.com/52034228/59842658-a58a6f00-935f-11e9-86d2-69d45bac7821.png)

**Log Trace**
![image](https://user-images.githubusercontent.com/52034228/59844052-935e0000-9362-11e9-8eb5-0582577135b1.png)

**If I remove the file called "kissa" and re-run the test, everything works fine + there is no more that weird file POST anymore:**

![image](https://user-images.githubusercontent.com/52034228/59844517-c2c13c80-9363-11e9-963f-f72166894ef3.png)

**Versions**
Chrome: 75.0.3770.100 and chrome webdriver 75 for it
selenium-server-standalone-3.141.59.jar
![image](https://user-images.githubusercontent.com/52034228/59844177-dc15b900-9362-11e9-9613-bf5a6517c080.png)
OS: Windows & Linux

**Comment by aaltat** [comment]
What you see is done by Selenium and is by design in Selenium. But why it actually happens, needs some explanation.

Selenium has a single API method, [send_keys](https://selenium-python.readthedocs.io/api.html#selenium.webdriver.remote.webelement.WebElement.send_keys), to send text to different kinds of `input` elements. Also file upload is an `input` element, although generally in browsers it looks like a button. If the browser happens to be running trough [Selenium grid](https://www.seleniumhq.org/docs/07_selenium_grid.jsp), example in SeleniumLibrary `Open Browser` keyword the `remote_url` is defined, then the Selenium will automatically transfer the file to the node where the browser is actually running, so that the file upload can succeed. The file is usually transferred to a temporary folder in the node.

Therefore if your string happens to match a file, by Python `os.path.isfile` and you are using Selenium Grid, then Selenium will decide that you are uploading a file and do the magic behind the scenes. Also there is not a simple way to disable the automatic file transfer. All the above is done by Selenium and not by SeleniumLibrary.

This Selenium feature is documented in the [Choose File](http://robotframework.org/SeleniumLibrary/SeleniumLibrary.html#Choose%20File) keyword. But for users, connectin the two keywords and understanding that the two keywords use the same underlying Selenium method is not realistic assumption. This makes me think should we update also the `Input Text` and `Input Password` keywords documentation to mention this corner case. Any good suggestions, how the documentation could be enhanced?

**Comment by pekkaklarck** [comment]
Could the affected keywords simply fail with a clear message if there exists a file matching whatever text is used?

**Comment by Perza-e** [comment]
Thanks for the clarification. Pekka's suggestion sounds good. Even if there would be documentation, it can be really time consuming to find the issue because the keyword itself won't fail. However, could it cause issue in performance if we loop through each variable and file every time the Input Text keyword is executed because the var value can change during the test run + user can create new files during the test run.

It took me nearly 3 hours to actually find the root cause of this problem. It actually happened on our Linux test automation server in which all our tests are run. We have multiple users and all of us have cloned the test repository under /home/username/testrepo/

My colleague had this issue in which a 3 hour long test was failing when submitting a specific web page. The Input Text keyword just passed fine even that was the guilty one. I had no issues running the same test case with same version of test repository so we had to start taking a video capture and enable trace log level in order to see what is going wrong and where.

We have hundreds of variables in our environment during a test run so it's hard to avoid this issue. We have some kind of hack in our env so that we don't need to know the path of specific resource nor test file. We can just import it by writing it's name even the file is located under different sub directory. So my colleague had a file called 'tunnel' in the root of his test repository and the actual test file was under testrepo/cases/platform/testcase.robot. This robot file uses a variable which value is 'tunnel' at specific step in the test case so that lead to this issue.

**Comment by aaltat** [comment]
Making keyword fail, if the argument points to a file with grid, may be backwards incompatible change. The `Choose File` file has feature, if the argument doesn't point to a file, keyword will fail. But if user wants to upload something else, like a folder, many files or something else which doesn't not work with `Choose File` the general recommendation, also from me, has been: Use `Input Text`.

Now when I quickly looked at the Selenium code, it looks like that Selenium also supports [files only](https://github.com/SeleniumHQ/selenium/blob/11c25d75bd7ed22e6172d6a2a795a1d195fb0875/py/selenium/webdriver/remote/file_detector.py#L47) with the grid and therefore changing the functionality may be possible. But in any case it needs careful testing and most likely it needs also testing with the Selenium grid in the Travis, separate issue here #1282 . Having grid support in acceptance test is not overly complicated or magical work, it's just work which needs some thinking how it can be done and then doing it. Generally it would be useful feature to have in the SeleniumLibrary acceptance test, because I have needed to have that functionality for few times already, but have not found time to do it. Would someone be interested for contributing?

**Comment by pekkaklarck** [comment]
Failing if text points to a file should only be added to kws such as Input Text, not to Choose File. Shouldn't cause any noticeable slowdown.

**Comment by aaltat** [comment]
Changing the functionality in that way sounds pretty good idea. `Choose File` would not perform any checking does the file, folder or whatever exist. It would blindly call Selenium `send_keys` and let the user worry about correctness of the argument.

Also it would be good idea change `Input Text` and `Input Password` to fail if the if Selenium grid is used and argument points to a file.

**Comment by aaltat** [comment]
I did do some research and it's also possible to overwrite the Selenium [file decorator](https://github.com/SeleniumHQ/selenium/blob/master/py/selenium/webdriver/remote/webdriver.py#L132) when browser is opened. Need to think is there a straight forward way to hook back from Selenium to the SeleniumLibrary and ask what it should do.

**Comment by aaltat** [comment]
The #1282 is now done and it is possible to use grid to run tests.

**Comment by aaltat** [comment]
I did split this issue to two different issues and therefore I am closing this issue.

**Comment by aaltat** [comment]
@Perza-e this problem should be now fixed in the master branch, by #1417 The `Input Text` and the `Input Password` should not accidentally do file upload. And only `Choose File` should do the file upload.

Would it be possible you to try it out and provide feedback?

**Comment by Perza-e** [comment]
@aaltat, I have now run the same test on Windows.
- Problem reproduces with old Selenium
- After installing the latest master build, the problem got fixed.

Good job and thank you for fixing it!

I cannot try this out with Linux because we got old Selenium running in a production environment in which I encountered with this issue for the first time. Upgrading Selenium would deprecate lots of keywords so I'll need to do quite much maintenance before.
